### PR TITLE
DEP: lib: Remove the deprecated financial functions.

### PR DESCRIPTION
**Pull request by WarrenWeckesser**
As explained in NEP 32, the financial functions are to be removed
from version 1.20.

### CODE REVIEW

**Comment by seberg** [comment]
Happy to do this. We could also go through an additional `VisibleDeprecationWarning` cycle, but I suppose, we always have the option to revert the removal and do that instead...

Lets sign it off in the meeting today (technically, its already accepted so...).

**Comment by eric-wieser** [comment]
Another option: replace the functions with stubs that just raise `RuntimeError`, just so that users get a better message than `AttributeError`

**Comment by seberg** [comment]
Hmm, the error idea is indeed maybe a bit gentler and only leaves like 50 lines of code probably. Too bad we can't put it directly on `numpy.financial` yet.

File: numpy/__init__.py [hunk 1]
```diff
@@ -228,6 +241,13 @@
         # module level getattr is only supported in 3.7 onwards
         # https://www.python.org/dev/peps/pep-0562/
         def __getattr__(attr):
+            # Raise AttributeError for expired attributes
+            try:
+                msg = __expired_attrs__[attr]
+                raise AttributeError(msg)
+            except KeyError:
+                pass
```
Comment (eric-wieser) [inline numpy/__init__.py:249]: nit for consistency with the code below:
```suggestion
            except KeyError:
                pass
            else:
                raise AttributeError(msg)
```

**Review by eric-wieser — APPROVED** [review]
Good thinking on using `__getattr__`. Maybe worth a test for this - checking that NEP 32 appears in the error message for just `np.fv` would be plenty.

**Comment by WarrenWeckesser** [comment]
More than one person suggested using the module `__getattr__` function to handle the error message.  I pushed a commit that implements this by adding `__expired_attributes__` in `numpy/__init__.py`.  It is handled similar to `__deprecated_attributes__`, but raises an `AttributeError` with an informative message instead of a warning.

For example,
```
In [1]: import numpy as np                                                                           

In [2]: np.npv                                                                                       
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-2-ab016edb79d9> in <module>
----> 1 np.npv

~/mc38/lib/python3.8/site-packages/numpy-1.20.0.dev0+2558224-py3.8-macosx-10.9-x86_64.egg/numpy/__init__.py in __getattr__(attr)
    245             try:
    246                 msg = __expired_attrs__[attr]
--> 247                 raise AttributeError(msg)
    248             except KeyError:
    249                 pass

AttributeError: In accordance with NEP 32, the function npv was removed from NumPy version 1.20.  A replacement for this function is available in the numpy_financial library: https://pypi.org/project/numpy-financial
```

In Python 3.6, a plain `AttributeError` will be raised, but we will likely drop (official) support for 3.6 in the 1.20 release, so maybe that's OK.

This adds a bit more overhead when accessing the `numpy` namespace, so perhaps this isn't something we want to do.

**Comment by seberg** [comment]
@WarrenWeckesser thanks, it doesn't actually add additional overhead. The `__getattr__` is only called if the item is not already found in the module dictionary. Or yes, it adds overhead, but only when attribute lookup fails :).

**Comment by WarrenWeckesser** [comment]
@seberg, thanks, that's good to know.

File: numpy/lib/tests/test_financial_expired.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import pytest
+import numpy as np
+
+
+def test_financial_expired():
+    with pytest.raises(AttributeError, match='NEP 32'):
+        np.fv
```
Comment (seberg) [inline numpy/lib/tests/test_financial_expired.py:7]: Maybe just put this into `numpy/tests/test_public_api.py`? Although, that file is mostly complex tests, so I don't care much.

**Review by seberg — APPROVED** [review]

File: numpy/lib/tests/test_financial_expired.py (hunk 2 above)
Comment (WarrenWeckesser) [reply numpy/lib/tests/test_financial_expired.py:7]: I thought about that, but `test_public_api.py` seems to have a specific purpose in maintaing the "standard" numpy API.  This new test doesn't seem to fit there.  It didn't fit well anywhere, so I just put it into its own file.

File: numpy/__init__.py (hunk 1 above)
Comment (WarrenWeckesser) [reply numpy/__init__.py]: Fixed in an update.

File: numpy/lib/tests/test_financial_expired.py (hunk 2 above)
Comment (seberg) [reply numpy/lib/tests/test_financial_expired.py:12]: Fair enough :).

**Comment by seberg** [comment]
Thanks, Warren, pretty amazing to see this actually finished! Now I just wonder how much fallout we will see ;).

**Pull request merged**